Fixes part of #21643: Checking CLA sign functionality moved from oppiabot to github actions.

Ashu463 opened the pull request:
## Overview

1. This PR fixes part of #21643 
2. This PR does the following: This PR fixes part of issue#21643, i.e. checking CLA sign through github actions. 
## Essential Checklist

Please follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).

- [X] I have linked the issue that this PR fixes in the "Development" section of the sidebar.
- [X] I have checked the "Files Changed" tab and confirmed that the changes are what I want to make.
- [ ] I have written tests for my code.
- [X] The **PR title** starts with "Fix #bugnum: " or "Fix part of #bugnum: ...", followed by a short, clear summary of the changes.
- [ ] I have assigned the correct reviewers to this PR (or will leave a comment with the phrase "@{{reviewer_username}} PTAL" if I can't assign them directly).

## Proof that changes are correct
TBD: Add screenshot of passing CI checks.

## PR Pointers

- Never force push! If you do, your PR will be closed.
- To reply to reviewers, follow these instructions: https://github.com/oppia/oppia/wiki/Make-a-pull-request#step-5-address-review-comments-until-all-reviewers-approve
- Some e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).
- See the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers) for what code owners will expect.

Ashu463 committed:
feat: added CI workflow for checking signing CLA

Ashu463 committed:
fix: lint fixes

Ashu463 committed:
fix: added authetication of spreadsheet

Ashu463 committed:
fix: updated workflow

Ashu463 commented:
Hey @U8NWXD we need to make sure that credentials present in oppiabot repository, must be present here in oppia/oppia repo too. I urge to pls check it once, if not then do create with same values present in oppiabot repo. 
Thanks.

Ashu463 committed:
Merge branch 'develop' into issue#21643-cla

seanlip commented:
@Ashu463 Please see https://github.com/oppia/oppia/issues/21643#issuecomment-2828510674. There are no secrets in the oppiabot repo, they are all on this repo.

Ashu463 committed:
fix: updated CI workflow

Ashu463 commented:
Hey @U8NWXD and @seanlip here is the video demo of checking CLA signed over issue or not. 
Here is the link of CI workflow which I showed in video: https://github.com/Ashu463/oppia-test-repo/actions/runs/14761112461/job/41441771521. 

One thing which I'm assuming is the JSON structure of SHEETS_CRED would be like: 
`
{
  "client_id": 
  "client_secret": 
  "redirect_uris":
  "refresh_token": 
}
`

https://github.com/user-attachments/assets/ce5fbcd9-1dcd-40fe-968d-4d40b1bfe472

PTAL once. 
Thanks.

seanlip commented:
@jayam04 Could you PTAL at this PR for codeowner training? Thanks.

jayam04 commented on .github/workflows/check_cla_assigned.yml:
```diff
@@ -0,0 +1,155 @@
+  check-cla:
+    runs-on: ubuntu-24.04
+    permissions:
+      pull-requests: write
```
@Ashu463 Does this action also run on PR? Referring to on_issue_assigned, it seems we are only working on Issue Assigned. If the later is true, do we need write permission on pull-requests?

Ashu463 replied:
CLA check should only work with issues, sorry that's my bad actually that I forgot to remove while sending it for review. Removed it.

Ashu463 replied:
Actually I was thinking to check CLA signing on PRs as well with the same workflow. Asked for your suggestion below.

jayam04 commented on .github/workflows/check_cla_assigned.yml:
```diff
@@ -0,0 +1,155 @@
+      - name: Install Dependencies
+        run: npm install googleapis google-auth-library
+
+      - name: Check CLA signature
```
@Ashu463 We are also checking if CLA is signed when `on_pr_opened` event occurs. If this event is not covered by this workflow, wouldn't it be better to have common action to check if CLA is signed?

Ashu463 replied:
Hmm, that's what I wanna discuss. So should we proceed the solution for checking CLA on both issues and PRs on a single workflow or seprate them? And yeah I'm in favour of making a common github action for both of them. WDYT?

jayam04 commented on .github/workflows/check_cla_assigned.yml:
```diff
@@ -0,0 +1,155 @@
+              
+              if(!response) {
+                console.log("No response from Google Sheets API");
+                core.setOutput('has-signed-cla', 'false');
+                return;
```
Can we add context due to which CLA isn't verified. ie Due to no response, and use the same when commenting. For example, "Couldn't verify CLA, as no response received from Google Sheets API".

Ashu463 replied:
Actually I'm logging things in the workflow. Are you saying to comment failure of verification on issues/PRs as well?

jayam04 commented on .github/workflows/check_cla_assigned.yml:
```diff
@@ -0,0 +1,155 @@
+            });
+            console.log(`Removed assignee ${assignee.login} from issue #${issue.number} due to unsigned CLA`);
+            
+      - name: Report failure
+        if: failure()
+        uses: actions/github-script@v6
```
We already have `send-webhook-notification` github-action, can we use that?

Ashu463 replied:
Hmm sure, synced this workflow with existing ones.

jayam04 commented:
@seanlip Done!

jayam04 commented:
@Ashu463 Thanks. Left few comments, PTAL.

Ashu463 committed:
fix: first iteration changes

Ashu463 commented:
Sorry @jayam04 I forgot to assign you, pls respond to my cmnts. Also, I created CLA check for PR as well in the same workflow, now I'm testing it on my side, hopefully you'll get it by late night. 
Thanks

Ashu463 committed:
fix: added CLA check feat for PRs as well

seanlip commented:
@Ashu463 Generally we should have consistent ways of handling things. If we already have a mechanism for this in the other PRs then let's follow that same mechanism here. Thanks.

jayam04 commented:
> Hey @jayam04 and @seanlip I found one thing, PRs couldn't access repository secrets and hence this workflow couldn't check CLA for PRs; this is explained well in this [article](https://michaelheap.com/access-secrets-from-forks/#:~:text=GitHub%20prevents%20PRs%20from%20forks,use%20them%20for%20malicious%20purposes.) Now it's solution is either give permissions to all users for accessing secrets in PR (which compromises the security) or to use github applications as we discussed previously with my all PRs. WDYT, how should I proceed? Thanks.

@Ashu463 I really thought why aren't we using GitHub Application on the first place, because I think they are more maintainable. We already have one app that comments when changes are deployed to test server ([here](https://github.com/apps/oppia-github-app)). However, as @Mayank77maruti has already implemented `on_pr_comment` in https://github.com/oppia/oppia/pull/21705 with github workflows, the former is okay too. 

For the first method, we definitely can't share secrets will all users, but can you give me more context what we are trying to do? We can comment in PR right, the issue is we need google secrets to access CLA agreement sheet. If I am right and if it's okay to share who signed CLA publicly, we can simply create a webhook in Google Sheets using Apps Scripts and access CLA agreements using a webhook URL.

jayam04 commented:
> > Hey @jayam04 and @seanlip I found one thing, PRs couldn't access repository secrets and hence this workflow couldn't check CLA for PRs; this is explained well in this [article](https://michaelheap.com/access-secrets-from-forks/#:~:text=GitHub%20prevents%20PRs%20from%20forks,use%20them%20for%20malicious%20purposes.) Now it's solution is either give permissions to all users for accessing secrets in PR (which compromises the security) or to use github applications as we discussed previously with my all PRs. WDYT, how should I proceed? Thanks.
> 
> @Ashu463 I really thought why aren't we using GitHub Application on the first place, because I think they are more maintainable. We already have one app that comments when changes are deployed to test server ([here](https://github.com/apps/oppia-github-app)). However, as @Mayank77maruti has already implemented `on_pr_comment` in #21705 with github workflows, the former is okay too.
> 
> For the first method, we definitely can't share secrets will all users, but can you give me more context what we are trying to do? We can comment in PR right, the issue is we need google secrets to access CLA agreement sheet. If I am right and if it's okay to share who signed CLA publicly, we can simply create a webhook in Google Sheets using Apps Scripts and access CLA agreements using a webhook URL.

/cc @seanlip

seanlip commented:
I think the issue that @Ashu463 is talking about is just that his PR needs access to secrets on oppia/oppia (to test whether it works), right?

If so, how have we solved this problem in the past? Let's follow that solution.

Ashu463 commented:
Hi @seanlip sorry for the delay, actually I was testing solution on my side before proposing it to you. 
- How CLA action is handled for PRs till now: Actually we're allowing `pull_requests_target` in place `pull_requests` to hit this workflow. In github, workflow trigerred by pull_request_target have write permissions on target repository not on the fork one. This whole thing is explained in this blog: https://securitylab.github.com/resources/github-actions-preventing-pwn-requests/ And that's why I'd updated workflow with target pull_request_target.
- Now the second thing I did is rather than passing SHEETS_CRED as JSON object, I passed it as encoded base64 string. Coz I read it somewhere that workflow do face some issues while fetching bulky JSON secrets. Thus I'm in favour to use base64 method, which worked on my test repository as shown in demo. 

Now for @jayam04 questions: Actually I wanna move a step-by-step approach i.e. first create a simple github action and hence move it to github application solution, I have that in my mind as I used it on my inactive issue PR as well. Now things which we're trying to do: 
- We wanna check is CLA signed by issue assignee or PR author or not? 
- If yes, then nothing will be happened on issue threads and CI check will pass in PRs. 
- But if not, then we comment and tell the end developer that he/she didn't signed CLA do this and that to sign. 

Umm, I don't think it's fine to public CLA signed user names, else it would have been done already by the one who made this oppiabot CLA action thing. Maybe @seanlip could you tell us more about it.  
WDYT guys? What should I do, should I carry on with this or move towards github application method?


https://github.com/user-attachments/assets/192162e1-157e-4165-95f8-7b490e218db4

Ashu463 commented:
Hey @seanlip should I move my solution to use base64 string in place of sheets cred JSON?

Ashu463 committed:
fix: typos

seanlip commented:
@Ashu463:

- Use a GitHub app, as you did with your previous PR that was approved.
- The JSON is not bulky and should be able to be used directly. I don't see the point of more conversions unless you've verified that using the JSON directly is actually causing an issue.
- We want to verify that anyone who opens a PR has signed the CLA. If not, close the PR, explain why you closed it, and tell the developer what to do about it. There is no requirement on issues.
- I don't know what you mean by "public CLA signed user names". Please clarify, thanks.

Ashu463 commented:
Thanks @seanlip for confirming, I'll update the solution and share demo in upcoming 48 hrs.

Ashu463 committed:
fix: updated the workflow to close the PR as well

Ashu463 commented:
@seanlip and @jayam04 I don't think we need to setup github app for this workflow, since this doesn't need any PAT which have any security issues. BTW, the issues which I was facing at that time is resolved and it doesn't need to accept encoded base64 version of SHEETS_CRED JSON. It all works well on my side, have a look on video: 

For PR: 

https://github.com/user-attachments/assets/a6cf3b29-4752-4b61-8fbf-11c025a020ba

For Issues: 

https://github.com/user-attachments/assets/300503d8-4211-4dd4-bfb4-ef2fbd97fd54

Ashu463 commented:
Friendly Ping @seanlip @jayam04

seanlip commented:
Hi @Ashu463, thanks for the ping. For discussion of GitHub  Actions workflows could you please collaborate with @U8NWXD primarily? I will assign him to this PR, you might want to do that for future conversations too. Thanks!

Ashu463 commented:
@seanlip yeah sure, @U8NWXD could you PTAL at the same? Gist is: Sean asked me to enclose the logic in the github application, but I think we won't need it, because I'm using /gh-script@v6 which doesn't require any GITHUB_TOKEN to be passed along with the workflow. Could you share your views over it? 
Thanks.

U8NWXD commented:
I agree @Ashu463 I don't think a github app is needed here. A GitHub app lets us avoid using PATs when we need access to repos beyond the one the workflow is running in. In this case, we only need access to the repo that holds the workflow (oppia/oppia), so the default GITHUB_TOKEN should be sufficient (and is actually preferable since it is ephemeral)

U8NWXD commented on .github/workflows/check_cla_assigned.yml:
```diff
@@ -0,0 +1,176 @@
+name: Check Issue Assignee CLA
```
Doesn't this workflow operate on both issues and PRs?

Ashu463 replied:
Yeah, I club them because of similarity in their parent work purpose.

U8NWXD replied:
But the workflow title only mentions issues. The workflow operates on both issues and PRs, and the title should reflect this

Ashu463 replied:
Got it, updated that with `Check CLA Signature for Issues and Pull Requests`.

U8NWXD commented on .github/workflows/check_cla_assigned.yml:
```diff
@@ -0,0 +1,176 @@
+  issues:
+    types:
+      - assigned
+permissions: read-all
```
top-level permissions should be minimal (i.e. `{}`)

Ashu463 replied:
Done, but later it would get stuck in CI checks where permissions are tested.

U8NWXD replied:
not sure what you mean. could you elaborate?

Ashu463 replied:
That's what I'm talking about: https://github.com/oppia/oppia/actions/runs/15982187098/job/45078880463?pr=22456, this shard test are written in format where they necessarily check top level permissions as read-all.

U8NWXD replied:
Ahh okay; in that case let's leave the permissions as `read-all`

Ashu463 replied:
Fine, made it read-all now.

U8NWXD commented on .github/workflows/check_cla_assigned.yml:
```diff
@@ -0,0 +1,176 @@
+        id: check-cla
+        uses: actions/github-script@v6
+        env:
+          SHEETS_CRED: '${{ secrets.SHEETS_CRED }}'
+          SPREADSHEET_ID: '${{ secrets.SPREADSHEET_ID }}'
```
do these already exist? If not, can you give them more descriptive names? (i.e. should specify what the sheets are for)

Ashu463 replied:
I used the exact same names given in oppiabot repository.

U8NWXD commented on .github/workflows/check_cla_assigned.yml:
```diff
@@ -0,0 +1,176 @@
+                console.log("SHEETS_CRED parsed successfully");
+              } catch (parseError) {
+                console.error("Failed to parse SHEETS_CRED:", parseError.message);
+                console.log("First 10 chars:", sheetsCredRaw.substring(0, 10) + "...");
```
we should never be logging tokens, even in part, in case we accidentally put a different token in SHEETS_CRED. If sensitive info from this credential has been written to logs during testing, we should revoke it and switch to a new token

If we avoid printing the token here, I think you can drop the try-catch too

Ashu463 replied:
I'm really sorry for that, actually earlier I was getting errors while fetching these credentials, removed it now. Yeah we could, actually I added try catch for same purpose.

U8NWXD replied:
have production tokens been written to logs during testing? I don't *think* the Oppia's production tokens should have been leaked since this workflow uses the `pull_request_target` event and so doesn't run in this PR (pull_request_target workflows run the code from develop, not the PR), but please double-check to be sure. Also, if you were using your own tokens for testing, these might have been leaked and should be revoked

Ashu463 replied:
I don't think so cause of the same reason of pull_request_target that runs on develop. And in past I hadn't created any workflow where oppia's secrets get logged out. As far as my credentials are concerned, luckily github won't log any secret information to workflow even if they are bound to print in the workflows, it just print a bunch of asterisks in a row.

U8NWXD commented on .github/workflows/check_cla_assigned.yml:
```diff
@@ -0,0 +1,176 @@
+              if (!sheetsCredJson.installed || !sheetsCredJson.installed.client_id || !sheetsCredJson.installed.client_secret || !sheetsCredJson.refresh_token) {
+                console.error("SHEETS_CRED is missing required credential fields");
+                core.warning("SHEETS_CRED is missing required credential fields");
+                core.setOutput('has-signed-cla', 'true');
```
is this "fail open" policy desired? CLA signing has legal implications for Oppia, so we might want to fail closed (i.e. default to blocking PRs if we can't confirm that the CLA has been signed)

btw it looks like oppiabot also fails open

/cc @seanlip

seanlip replied:
@U8NWXD Confirming that not "failing open" is the correct thing to do, i.e. if we cannot confirm signing then this check should fail, not pass. Thanks for catching this!

Also, if there is some way to convey "not true, not false, there's some error in the process" (i.e. "unable to verify") then that would probably be semantically better here than just saying "false". In one case we can say "an error occurred, please contact an admin" and in the other case we can say "you haven't signed it, please do so". In fact in the latter case it is fine to close their PR and ask them to reopen once they've signed, but we wouldn't want to do that in the former.

Ashu463 replied:
That's what I wanna discuss, thanks for raising it. Marked false.

U8NWXD replied:
How about we keep the same `true` vs `false` output values, but if we can't tell whether the CLA has been signed, we throw an error and fail the workflow? That will block PR merges (so long as a passing workflow is required for PR merging) but not close the PR. It also takes advantage of GitHub's built-in error handling rather than building something new, and throwing an error correctly conveys that something went wrong

Ashu463 replied:
Hmm, that's great way to follow. Now I'm logging the error and raising up the warning over the step and hence throwing error to fail the workflow. PTAL there.

U8NWXD commented on .github/workflows/check_cla_assigned.yml:
```diff
@@ -0,0 +1,176 @@
+              });
+
+              const sheets = google.sheets({ version: 'v4', auth: oauth2Client });
+              console.log(`Fetching data from spreadsheet ID: ${process.env.SPREADSHEET_ID}`);
```
Let's avoid printing out anything from `secrets` in case someone makes a mistake and puts something sensitive in the `SPREADSHEET_ID` secret

Ashu463 replied:
That's my bad, similar reason as said earlier.

U8NWXD commented on .github/workflows/check_cla_assigned.yml:
```diff
@@ -0,0 +1,176 @@
+            } catch (error) {
+              console.error("Error during CLA verification:", error);
+              core.warning(`CLA check failed: ${error.message}`);
+              // Don't fail the workflow on error, default to allowing the contribution
+              core.setOutput('has-signed-cla', 'true');
```
ditto: do we really want to fail open? /cc @seanlip

seanlip replied:
Confirming that we should not fail open.

Ashu463 replied:
Ditto: Marked false

U8NWXD commented on .github/workflows/check_cla_assigned.yml:
```diff
@@ -0,0 +1,176 @@
+          script: |
+            // Determine if we're handling a PR or an issue
+            const isPullRequest = !!context.payload.pull_request;
+            const isIssue = !!context.payload.issue;
```
```suggestion
            const isPullRequest = Boolean(context.payload.pull_request);
            const isIssue = Boolean(context.payload.issue);
```

I think this is more readable

Ashu463 replied:
Hmm, done.

U8NWXD commented on .github/workflows/check_cla_assigned.yml:
```diff
@@ -0,0 +1,176 @@
+              }
+
+              const linkToCla = '[here](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#setting-things-up)';
+              commentBody = `Hi @${assignee.login}, you need to sign the CLA before you can get assigned to issues. Follow the instructions ${linkToCla} to get started. I am unassigning you for now, feel free to assign yourself once you have signed the CLA. Thanks!`;
```
```suggestion
              commentBody = `Hi @${assignee.login}, you need to sign the CLA before you can get assigned to issues. Follow the instructions ${linkToCla} to get started. I am unassigning you for now; feel free to assign yourself once you have signed the CLA. Thanks!`;
```

Ashu463 replied:
Done.

U8NWXD commented on .github/workflows/check_cla_assigned.yml:
```diff
@@ -0,0 +1,176 @@
+            // Determine if we're handling a PR or an issue
+            const isPullRequest = !!context.payload.pull_request;
+            const isIssue = !!context.payload.issue;
+            const assignee = context.payload.assignee;
```
why are we grabbing the assignee differently here than we did to find `USERNAME` above? Is there a chance those two might be different, and if so, how should that be handled?

Ashu463 replied:
It's fine to eliminate assignee and take use of username in both the steps of workflow. I'd tested it on my test-repository and it works fine for me. Will share you the link of PRs/issues in threads below.

U8NWXD commented on .github/workflows/check_cla_assigned.yml:
```diff
@@ -0,0 +1,176 @@
+
+              console.log(`Closed and commented on PR #${pr.number} due to unsigned CLA`);
+            } else {
+              core.setFailed('Unknown event type or missing context');
```
can you give more detail here about what went wrong?

Ashu463 replied:
Yeah sure, I logged contextInfo object containing username, event and payload sent to that workflow step.

U8NWXD commented:
Please add proof of correctness. Note that passing CI checks are insufficient here. You should show in a testing repo that the action works as expected for both issues and PRs and for users who have and haven't signed the CLA

Ashu463 committed:
fix: iteration#1

Ashu463 commented:
Hey @U8NWXD I addressed all your comments, updated workflow as per your directions PTAL. Here are the proofs from my test repository: 
- CLA signed guy opened a PR: https://github.com/Ashu463/oppia-test-repo/pull/10
- PR opened by one who hadn't signed CLA: https://github.com/Ashu463/oppia-test-repo/pull/9
- There were some issues where I tried to assign CLA guy and non-cla one, you may check yourself I'd added you as a collaborator on `Ashu463/oppia-test-repo`. Here is link: https://github.com/Ashu463/oppia-test-repo/issues. 

Thanks.

Ashu463 committed:
Merge branch 'develop' into issue#21643-cla

Ashu463 reviewed (commented):
PTAL now.

U8NWXD commented:
@jayam04 could you elaborate on how moving these workflows into `/scripts/` would work? Workflows have to be under `/.github/workflows/` for GitHub to find them, so would we need to have code in `/scripts/` that is referenced by workflows in `/.github/workflows/`? If so, I worry that this will introduce more complexity than it's worth

U8NWXD commented on .github/workflows/check_cla_assigned.yml:
```diff
@@ -0,0 +1,174 @@
+      pull-requests: write
+      issues: write
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v4
```
do we really need to checkout the repository? I don't think we actually do anything in this workflow that needs the repo checked-out

Ashu463 replied:
Yeah, great catch! we could definitely remove this step out of the workflow. And I think we won't need this step in any of the workflow during migration (in most of).

U8NWXD replied:
not done

Ashu463 replied:
I'm sorry, I forgot that. Removed it now.

Ashu463 committed:
fix: iteration#2

U8NWXD commented on .github/workflows/check_cla_assigned.yml:
```diff
@@ -0,0 +1,174 @@
+              if (!sheetsCredJson.installed || !sheetsCredJson.installed.client_id || !sheetsCredJson.installed.client_secret || !sheetsCredJson.refresh_token) {
+                console.error("CLIENT_SECRET is missing required credential fields");
+                core.warning("CLIENT_SECRET is missing required credential fields");
+                core.setOutput('has-signed-cla', 'false');
```
this could still be confusing since we don't *know* that the CLA hasn't been signed. Instead, how about we don't set `has-signed-cla` at all and we change `if: steps.check-cla.outputs.has-signed-cla == 'false'` below to `if: steps.check-cla.outputs.has-signed-cla != 'true'` so that if anything goes wrong, we still fail closed?

Ashu463 replied:
Hmm, sounds good to me, and now I should raise error from `core.setFailed` and update below step condition with != true.

U8NWXD commented on .github/workflows/check_cla_assigned.yml:
```diff
@@ -0,0 +1,174 @@
+
+              if (!sheetsCredJson.installed || !sheetsCredJson.installed.client_id || !sheetsCredJson.installed.client_secret || !sheetsCredJson.refresh_token) {
+                console.error("CLIENT_SECRET is missing required credential fields");
+                core.warning("CLIENT_SECRET is missing required credential fields");
```
I recommend using `core.setFailed` instead since that both sets an error annotation and ensures the workflow exits with a nonzero status code (i.e. fails).

ditto elsewhere

Ashu463 replied:
Thanks, I'd updated those part of workflow where it is supposed to be failed.

Ashu463 committed:
fix: updated workflow

jayam04 commented:
@Ashu463 I am not reviewing the PR. But my query was regarding using separate script and @U8NWXD have answered it. Thanks, @U8NWXD I think let's stick to this then.

LGTM!

Ashu463 commented:
Thanks @jayam04 for LGTM, @U8NWXD could we merge this PR?

Ashu463 committed:
Merge branch 'develop' into issue#21643-cla

Ashu463 committed:
Merge branch 'develop' into issue#21643-cla

Ashu463 committed:
Merge branch 'develop' into issue#21643-cla
